[Task] Make bundle migrations idempotent

mattamon opened the pull request:
<!--

Before working on a contribution, you must determine on which branch you need to work:
- Bug fix: choose the latest maintenance branch `10.5`
- Feature/Improvement: choose `11.x` 

> All bug fixes merged into the latest maintenance branch are also merged to the current dev branch (`11.x`) on a regular basis.

## Please make sure your PR complies with all of the following points: 
- [ ] Read and accept our [contributing guidelines](/CONTRIBUTING.md) before you submit a PR.
- [ ] Features need to be proper documented in `doc/` 
- [ ] Bugfixes need a short guide how to reproduce them -> target branch is the oldest supported maintenance branch, e.g. `10.5` (see Readme.md for the list of supported versions)
- [ ] Meet all coding standards (see PhpStan actions) 

**Don't submit a PR if it doesn't comply, it'll be closed without a comment!**
-->  
  

## Changes in this pull request  
Resolves #14340 

## Additional info  
Removing the settings store setting in down method

mattamon committed:
Add delete settings store setting in down method

mattamon committed:
Fix condition

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221228101109.php:
```diff
@@ -55,6 +55,9 @@ public function up(Schema $schema): void
 
     public function down(Schema $schema): void
     {
+        if (SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\CustomReportsBundle\\PimcoreCustomReportsBundle', 'pimcore')) {
```
the `up` here also misses the `get` to check if is already in the settingstore

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221228101109.php:
```diff
@@ -55,6 +55,9 @@ public function up(Schema $schema): void
+        if (SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\CustomReportsBundle\\PimcoreCustomReportsBundle', 'pimcore')) {
+            SettingsStore::delete('BUNDLE_INSTALLED__Pimcore\\Bundle\\CustomReportsBundle\\PimcoreCustomReportsBundle', 'pimcore');
+        }
         $this->write(sprintf('Please deactivate the %s manually in config/bundles.php', PimcoreCustomReportsBundle::class));
```
This could be a nice-to-have in every bundle migraton

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20221215071650.php:
```diff
@@ -45,6 +45,8 @@ public function up(Schema $schema): void
     public function down(Schema $schema): void
     {
-        // do nothing
+        if (SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\GlossaryBundle\\PimcoreGlossaryBundle', 'pimcore')) {
```
here too

kingjia90 commented on bundles/CoreBundle/src/Migrations/Version20230120111111.php:
```diff
@@ -54,6 +54,9 @@ public function up(Schema $schema): void
 
     public function down(Schema $schema): void
     {
+        if (SettingsStore::get('BUNDLE_INSTALLED__Pimcore\\Bundle\\ApplicationLoggerBundle\\PimcoreApplicationLoggerBundle', 'pimcore')) {
```
this `up` also

kingjia90 reviewed (changes requested):
The changes looks fine, but there are some inconsistencies of the merged migrations that may require some further changes.

mattamon committed:
Unify migrations

mattamon committed:
Merge branch '11.x' into 14340-make-bundle-migrations-idempotent

mattamon committed:
Adjust Personalization and GoogleMarketing

kingjia90 commented:
Thank you for fixing all that!

Just noticed some other problems, unrelated to this PR's change. 

- [x] ApplicationLogger: missing [permission category](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/ApplicationLoggerBundle/src/Installer.php#L80-L84) change + if valid, migration should also update the category
- [x] CustomReports: missing [permission category](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/CustomReportsBundle/src/Installer.php#L44-L49)+ if valid, migration should also update the category
- [x] FileExplorer: missing migration for [permission category ](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/FileExplorerBundle/src/Installer.php#L47)
- [x] GlossaryBundle: same
- [x] Google Marketing: same
- [x] Personalization: same
- [x] Seo: same
- [x] StaticRoute: same

at first glance, il could be that the other squad didn't know about the permission category change or they decided that it should stay in the core permissions, but if it's a valid input, i would appreciate if these change could be added to this "cross-bundle" PR

mattamon committed:
Adding permission updates to migrations and installers

mattamon commented:
> Thank you for fixing all that!
> 
> Just noticed some other problems, unrelated to this PR's change.
> 
>     * [x]  ApplicationLogger: missing [permission category](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/ApplicationLoggerBundle/src/Installer.php#L80-L84) change + if valid, migration should also update the category
> 
>     * [x]  CustomReports: missing [permission category](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/CustomReportsBundle/src/Installer.php#L44-L49)+ if valid, migration should also update the category
> 
>     * [x]  FileExplorer: missing migration for [permission category ](https://github.com/pimcore/pimcore/blob/1e9f6135c0d112f0ae4cb7fc247a12d778847b32/bundles/FileExplorerBundle/src/Installer.php#L47)
> 
>     * [x]  GlossaryBundle: same
> 
>     * [x]  Google Marketing: same
> 
>     * [x]  Personalization: same
> 
>     * [x]  Seo: same
> 
>     * [x]  StaticRoute: same
> 
> 
> at first glance, il could be that the other squad didn't know about the permission category change or they decided that it should stay in the core permissions, but if it's a valid input, i would appreciate if these change could be added to this "cross-bundle" PR

It is a totally valid input and I added the Permission Category where I could! 
Google Marketing -> there is a new permission like with XLIFF
FileExplorer -> there is no migration for this Bundle, we decided to not install this per default.

ApplicationLogger and CustomReports I added the category to the installer.

SimpleBackendSearch -> there is no permission for this one, and also no migration.
